Sort unstructured boundary conditions

andrewwinters5000 opened the pull request:
Improve the performance and type stability of the `calc_boundary_flux` of the unstructured solver. Addresses part of #542

ranocha commented on src/semidiscretization/semidiscretization_hyperbolic.jl:
```diff
@@ -81,22 +81,47 @@ end
+        indices_for_current_type = vcat(indices_for_current_type, temp_indices)
+      end
+    end
+    boundary_indices[j] = sort(indices_for_current_type)
```
```suggestion
    boundary_indices[j] = sort!(indices_for_current_type)
```
Minor performance optimization

ranocha commented on src/semidiscretization/semidiscretization_hyperbolic.jl:
```diff
@@ -81,22 +81,47 @@ end
+
+  # put everything together into a NamedTuple that has tuples with the (sorted) boundary type and
+  # its associated indices. Save a copy of the original dictionary for printing
+  ( sorted_boundary_information = tuple(zip(boundary_types, boundary_indices)...),
+    boundary_dictionary = boundary_conditions, )
```
The sorting you did above is already quite good, thanks :+1: However, i think it would be cleaner to create a new `struct` to hold the information you return here and use that for dispatch later. It could be something like
```julia
struct AndrewPicksSomeGreatNameHereSinceDontWantToDoThatRightNow{N, BCs<:NTuple{N, Any}}
  boundary_conditions::BCs
  boundary_indices::NTuple{N, Vector{Int}}
  boundary_dictionary::Dict{Symbol, Any}
end
```
Then, you can iterate over the tuples in a type-stable way using "lispy tuple programming" - search for that phrase in Trixi's source code, I'm sure I used it twice or so. That should remove the remaining allocations and make the `rhs!` stuff type stable and fast.

andrewwinters5000 replied:
This is a nice idea. It would also make the display `if` for the semidiscretization a bit cleaner. I just need to figure out where this struct would live. Would I need to make a new `semidiscretization` constructor? Or could I still pass this new struct as the general `boundary_conditions`argument?

ranocha replied:
You can let the `struct` live in the quad mesh solver file and construct it in `digest_boundary_conditions` as you do now.

andrewwinters5000 replied:
Groovy, I will work on it. My naive way already reduced memory allocations a bit but hopefully this struct way will also reduce the actual computation time

ranocha reviewed (changes requested):
Thanks for working on this! Please find below some comments. Feel free to ping me here or on Slack to discuss anything further.

sloede commented on src/solvers/dg_unstructured_quad/dg_2d.jl:
```diff
@@ -221,22 +221,27 @@ function calc_boundary_flux!(cache, t, boundary_condition::BoundaryConditionPeri
 
 
-function calc_boundary_flux!(cache, t, boundary_conditions,
+function calc_boundary_flux!(cache, t, boundary_conditions::NamedTuple,
```
```suggestion
function calc_boundary_flux!(cache, t, boundary_conditions,
```
I don't think the type annotation is required here, or is it?

andrewwinters5000 replied:
This probably stems from my own misunderstanding of type stability. I thought I had to tell the function the explicit type of an arguement to make it type stable. I assume from this comment that I am mistaken?

jlchan replied:
Yeah, annotation is usually just for specifying which function to dispatch. Less restrictive also means you can switch the type of `boundary_conditions` later.

andrewwinters5000 replied:
Noted. I removed the annotation

sloede reviewed (changes requested):
LGTM in general, with a minor comment. Once the other comments by @ranocha are resolved, this can be reviewed once more and then merged I think.

ranocha commented on src/solvers/dg_unstructured_quad/dg_2d.jl:
```diff
@@ -223,35 +223,65 @@ end
+# Iterate over tuples of boundary condition types and associated indices
+# in a type-stable way using "lispy tuple programming".
+function calc_boundary_flux_by_type!(cache, t, BCs::NTuple{N,Any},
+                                     BC_indices::NTuple{N,Vector{Int64}},
```
```suggestion
                                     BC_indices::NTuple{N,Vector{Int}},
```

ranocha reviewed (commented):
Looks already quite good. Just a minor comment and we can merge it from my side.

A general comment: Please consider using the GitHub feature to apply proposed changes directly. You would have saved both of us quite some time by doing that for https://github.com/trixi-framework/Trixi.jl/pull/570#discussion_r628171423 where I already fixed the problematic allocations I fixed in 875ad6ce7e55dc5bcd2aafbb9fc2cf29ec0b88c5

andrewwinters5000 commented:
> Looks already quite good. Just a minor comment and we can merge it from my side.
> 
> A general comment: Please consider using the GitHub feature to apply proposed changes directly. You would have saved both of us quite some time by doing that for [#570 (comment)](https://github.com/trixi-framework/Trixi.jl/pull/570#discussion_r628171423) where I already fixed the problematic allocations I fixed in [875ad6c](https://github.com/trixi-framework/Trixi.jl/commit/875ad6ce7e55dc5bcd2aafbb9fc2cf29ec0b88c5)

Okay, I will start using this feature. I used to like and combine the changes into a single commit to avoid cluttering the git log. But now that we use squash merge for PRs this is no longer an issue

ranocha commented:
> Okay, I will start using this feature. I used to like and combine the changes into a single commit to avoid cluttering the git log. But now that we use squash merge for PRs this is no longer an issue

You can go to the "Files changed" section at the top and select something like "Add suggestion to batch" for all of them. Then, you can commit them in one batch/commit.

andrewwinters5000 commented:
> You can go to the "Files changed" section at the top and select something like "Add suggestion to batch" for all of them. Then, you can commit them in one batch/commit.

Good to know. I will make sure I use this in the future

sloede commented on src/solvers/dg_unstructured_quad/dg_2d.jl:
```diff
@@ -223,35 +223,65 @@ end
+  # process the first boundary condition type
+  calc_boundary_flux!(cache, t, boundary_condition, boundary_condition_indices, mesh, equations, dg)
+
+  # recursively call this method with the unprocessed boundary types
+  calc_boundary_flux_by_type!(cache, t, remaining_boundary_conditions,
+                              remaining_boundary_condition_indices, mesh, equations, dg)
```
TBH, I don't really like these fancy lisp-style iterations unless explicitly required for performance reasons. I am not sure if this was benchmarked or not, but it makes the code harder to read and understand (IMHO, almost always does). No objections against using it here, but if we can get away without these constructs except for some extremely critical cases, I'd appreciate it.

efaulhaber commented on src/semidiscretization/semidiscretization_hyperbolic.jl:
```diff
@@ -81,22 +81,27 @@ end
-function digest_boundary_conditions(boundary_conditions::AbstractArray)
+# sort the boundary conditions from a dictionary and into tuples
+function digest_boundary_conditions(boundary_conditions::Dict, cache)
+  UnstructuredQuadSortedBoundaryTypes(boundary_conditions, cache)
```
I think I'll take the same approach. Maybe I can actually use @andrewwinters5000's code here as is.

sloede replied:
OK, if you already know how to possibly extend this for your purposes, that's good enough for now :relaxed:

ranocha replied:
I think we should use the "standard Trixi" approach: Implement one working version and extend/generalize it when the need arises (a second application).

efaulhaber commented:
The changes in this PR seem to add some more complexity. How does the performance compare to the stuff benchmarked by @ranocha in #573?

sloede commented:
It actually improves the performance, according to some benchmarks shown to me by @andrewwinters5000

ranocha commented:
The high-level overview: The first version implemented by @andrewwinters5000 payed the price of dynamic (runtime) dispatch (type instability) for every node on every boundary when at least two different BCs were used. My PR reduced that to runtime dispatch once per boundary. This PR fixed the underlying type instability completely, removing the cost of dynamic dispatch.

sloede commented:
I'm still surprised though that you can see the difference between the second and third versions. Intuitively, I would've thought this should be negligible.

ranocha commented:
> I'm still surprised though that you can see the difference between the second and third versions. Intuitively, I would've thought this should be negligible.

It had less impact, that's true. A bigger performance improvement came from

> A general comment: Please consider using the GitHub feature to apply proposed changes directly. You would have saved both of us quite some time by doing that for [#570 (comment)](https://github.com/trixi-framework/Trixi.jl/pull/570#discussion_r628171423) where I already fixed the problematic allocations I fixed in [875ad6c](https://github.com/trixi-framework/Trixi.jl/commit/875ad6ce7e55dc5bcd2aafbb9fc2cf29ec0b88c5)

Nevertheless, I think it's good to avoid type instabilities and train people to get used to some of these design choices.
